### ISSUE TITLE
Overview: Phase 2 Full Site Editing

### ISSUE THREAD

**Issue by epiqueras** [WordPress/gutenberg#17284]
*This issue is to serve as a live recap of the work being done to support full site editing in core.*

**The merged and pending final PRs are all here:** https://github.com/WordPress/gutenberg/pulls?q=is%3Apr+author%3Aepiqueras+project%3AWordPress%2Fgutenberg%2F13

Keep reading to understand how everything fits together.

# The Goals & Prototype

Most of the ideas of what full site editing should look like were born out of this prototype:

#16565.

### Dynamic Templates

The main breakthrough comes from how we decided to move forward with #4659. That PR from a long time ago, introduced a clever way of extending the traditional WordPress template hierarchy,

![image](https://user-images.githubusercontent.com/19157096/64048370-5f534800-cb26-11e9-985e-7df0d1f7c24b.png)

, with dynamic, user-editable templates which are just posts of a new custom post type. To summarize how it does this as briefly as possible, it hijacks the hierarchy lookup to stop at every level and pick a dynamic template, if there is one, over the traditional template file. Of course, since these templates are just posts, template files could also manually include their content.

### Template Editing

The goal of all this was to be able to, in parallel to the currently edited post, load these new dynamic template posts in the block editor and provide real full site editing. Editing that renders the entire site just how it would look like when previewing the post, and lets you edit and compose everything from reusable template parts like a header block or a sidebar, to the navigation menu, and of course, the actual post content. We also wanted users to be able to edit a template directly, separately from any post that might use it, and see a placeholder for post related content if needed. A priority in all this, albeit we almost gave up on it in multiple prototypes, was to have all of this work with a single block list in order to maintain the current seamless editing flow.

### Editor Modes

We appreciated that whatever new editing contexts we came up with, there would be trade offs for different users, so we explored how the editor could operate on different modes, each tailored to a specific goal or use case. We explored how a full site mode could render the post in its template, how we could provide focus modes to zero in on the post content, header, footer, or any other template part of the given template, and even tested a few more exotic modes like a design mode that turned on a layout debugger to fix alignment issues.

The prototype achieved all of these goals, but with lots of anti-patterns, and quick hacks. So, we stepped back for a bit and designed for the future.

# The Design & Implementation

### Handling Entities

In stepping around our code and discussing lots of different potential implementations. We realized that a lot of the complexity in the prototype came from trying to switch between editing one set of posts, e.g. some template parts and an actual post, to another set of posts. It was hard to keep things like edits, change detection, and undo levels, in sync. But, we already had an abstraction built for that, the core data store's entities. Unfortunately, most of the code in the editor had avoided using that abstraction, due to a lack of support for things like editing and history tracking, and instead opted to take the quicker route of directly interfacing with the API for regular posts. We needed to refactor this before we could move forward.

The following PRs took care of abstracting away and replacing all of the manual post handling in the editor into core data entities:

- #16814: Editor: Move auto-draft status change to the server.
- #16823: Core Data: Add util for minimally modifying items.
- #16867: Core Data: Add support for entity edits and undo history.
- #16903: Core Data: Add support for autosaving entities.
- #16932: Editor: Update the store to use Core Data entities.
- #17030: Core Data: Return updated record in `saveEntityRecord`.

### A Declarative API for Template Blocks

With the foundation ready, we needed to decide on an API for creating blocks that sync with different entities. So that for example, a block layout like the following could be achieved:

![image](https://user-images.githubusercontent.com/19157096/64049934-6761b680-cb2b-11e9-899b-c57fbf16d9e4.png)

#### Trials

There were quite a few ways to go about it. There was the possibility of extending the custom sources implementation (#16402) to also operate on inner blocks. But that only led us to discuss how unnecessarily complex custom sources were and mark them for refactoring (which the final version of the new API easily did). Among other notable approaches were:

- #17020: An algebraic effects inspired API for blocks to throw persistence effects that were handled by the nearest parent with a registered handler. This was nice, but unnecessarily complex, like custom sources.
- #17118: A simple entity provider with nested block editors. This was simple to implement, but splitting the block list broke a lot of design and implementation assumptions that peripheral editor components made.
- #17135: The best of the two previous approaches, a simple entity provider and controlled inner blocks instead of a block list split.

#### Final API

The final API ended up looking a lot like #17135.

It introduces a declarative and composable entity provider with dynamic entity type-kind-specific context creation:

```jsx
<EntityProvider kind="site" type="root">
  <EntityProvider kind="postType" type={ post.type } id={ post.id }>
    { children }
    <InnerBlocks />
  </EntityProvider>
</EntityProvider>
```

A few nice hooks for leveraging entities:

```js
const id = useEntityId( 'postType', 'post' );
const [ content, setContent ] = useEntityProp( 'postType', 'post', 'content' );
const [ isDirty, isSaving, save ] = useEntitySaving(
  'postType',
  'post',
  'content'
);
```

And controlled inner blocks:

```jsx
<InnerBlocks
  value={ blocks }
  onChange={ setBlocks }
  onInput={ setContent }
/>
```

These APIs were implemented and leveraged in the following PRs:

- #17153: Starts implementing the API and uses it to refactor custom sources away for a **20% typing performance speed up**.
- #17207: Implements the site title block.
- #17263: Implements the template mode and the post title and content blocks.

# Future Work

#17263 is the `HEAD` of this effort and contains all of the functionality behind the new Full Site Editing experiment in the experiments tab (#16626).

The framework changes and new APIs are done and now it's time to design and polish things like modes, mode switching, and save flows. We need to answer questions like:

#### Templates

- Where in the editor should we expose template selection for a post?
- Where do we expose templates for direct editing?
- What templates from the hierarchy do we want to expose to users?

#### Saving

- How do we show the user their edits are affecting multiple entities?
- Do we show a modal with checkboxes next to the affected entities when saving?
- Do we render inline save buttons in each block?

#### Modes

- What editor modes do we want to ship?
- How should we switch between them?

###### To be continued...
